File: 8/19/3.cpp
#include <string>
#include <vector>
#include <map>
#include <algorithm>
#include <iostream>

using namespace std;

template<template<typename> class P = less>


struct compare_pair_second{
    template<class T1, class T2> bool operator()(const pair<T1,T2>&left, const pair<T1,T2> & right){
        return P<T2>()(left.second,right.second);
    }


};



int prim(vector<vector<pair<int,int> > > g,int start){

    vector<int> indices;
    int ans = 0;

    indices.push_back(start);

    

    while(indices.size()<g.size()){

        int v1 = -1;
        int v2 = -1;
        int min = 9999999999;
        for(int i = 0; i < indices.size(); i++){
            if(count(indices.begin(),indices.end(),g[indices[i]][0].first) > 0){
                    g[indices[i]].erase(g[indices[i]].begin());
                    continue;
                }
        
            if(g[indices[i]].size() >0){
                if(min>=g[indices[i]][0].second){
                    v1 = indices[i];
                    v2 = g[indices[i]][0].first;
                    min = g[indices[i]][0].second;
                }
            }
        }

        
        indices.push_back(v2);

        
        ans+=g[v1][0].second;
        g[v1].erase(g[v1].begin());
        for(auto it = g[v2].begin() ; it != g[v2].end(); ++it){
            if(it->first == v1){
                g[v2].erase(it);
                break;
            }
        }


    }

    


    return ans;
}  



int solution(int n, vector<vector<int>> costs) {
    int answer = 0;

    vector<map<int,int> > graph(n);
    vector<vector<pair<int,int> > > gVec;

    if(n==1)
        return 0;

    for(int i = 0 ;  i < costs.size(); i++){
        graph[costs[i][0]][costs[i][1]] = costs[i][2];
        graph[costs[i][1]][costs[i][0]] = costs[i][2];
    }    

    
    for(int i = 0; i <n;i++){
        vector<pair<int,int> > tmpVec;
        for(auto iter = graph[i].begin(); iter!= graph[i].end(); ++iter){
            tmpVec.emplace_back(make_pair(iter->first,iter->second));
        }
        
        sort(tmpVec.begin(),tmpVec.end(),compare_pair_second<less>());
        gVec.push_back(tmpVec);
    }

    answer = prim(gVec,0);



    return answer;
}

File: 2021/4/22/3.py

def solution(n, s, a, b, fares):

    INF = float('inf')

    graph = [[INF for _ in range(n+1)] for _ in range(n+1)]

    dist = [[INF for _ in range(n+1)] for _ in range(n+1)]

    for fare in fares :
        graph[fare[0]][fare[1]] = fare[2]
        graph[fare[1]][fare[0]] = fare[2]

    for i in range(n+1) :
        for j in range(n+1) :
            dist[i][j] = graph[i][j]
        dist[i][i] = 0


    for k in range(n+1) :
        for i in range(n+1) :
            for j in range(n+1) :
                if dist[i][j] > dist[i][k] + dist[j][k] :
                    dist[i][j] = dist[i][k] + dist[j][k]

    answer = dist[s][a]+dist[s][b]
    for k in range(n+1) :
        if k == s :
            continue
        answer = min(answer,dist[s][k]+dist[k][a]+dist[k][b])


    return answer